admin/python/lsst/qserv/admin/nodePool.py

```diff
@@ -0,0 +1,160 @@
+_LOG = logging.getLogger(__name__)
+
+def cpu_count():
+    ''' Returns the number of CPUs in the system
+    '''
+    try:
+        return os.sysconf('SC_NPROCESSORS_ONLN')
+    except (ValueError, OSError, AttributeError):
+        return 1
+
+class NodePool(object):
+    """
+    Class providing tools to send ssh command to a set of machines.
+    Use system ssh.
+    """
+
+    def __init__(self, nodeAdmins=None, max_task=None):
+        """
+        Container of nodes which will be managed in parallel using ssh
+
+        @param nodeAdmins:  a list of NodeAdmin instances, each node has to be
+                            contactable via SSH
+        @param max_task:    number of parallel SSH command which will be run
+                            concurrently, equal to the number of cpu of the local
+                            machine by default
+        """
+
+        self.nodes = nodeAdmins
+
+        if max_task:
+            self._max_task = max_task
+        else:
+            self._max_task = cpu_count()
+
+    def execParallel(self, command, stdin=os.devnull):
+        """
+        Execute one SSH command in parallel on a set of nodes;
+        fork one process for each node.
+        Log of each command are stored in $CWD/parallel-cmd-log
+
+        @param command: command to run
+        @param stdin: file redirect to SSH client process standard input
+
+        @return the number of nodes in error
+        """
+        nodes_remaining = list(self.nodes)
+
+        if not command:
+            return  # empty list
+
+        logdir = "parallel-cmd-log"
+        try:
+            os.makedirs(logdir)
+        except OSError:
+            if not os.path.isdir(logdir):
+                raise
+
+        def logfile(name):
+            f = "{0}-{1}-{2}.txt".format(node.host, process_id, name)
+            return os.path.join(logdir, f)
+
+        def done(p):
+            return p.poll() is not None
+
+        def success(p):
+            return p.returncode == 0
+
+        max_task = cpu_count()
+        _LOG.info("nb node %s, max task: %s", len(nodes_remaining), max_task)
+        processes = []
+        process_id=0
+        nodes_failed = []
+
+        _LOG.info("Run command %s, stdin redirect to: %s", command, stdin)
+
+        while True:
+
+            # batch processes
+            while nodes_remaining and len(processes) < max_task:
+                process_id += 1
+                node = nodes_remaining.pop()
+
+                task = node.getSshCmd(command);
+                if _LOG.isEnabledFor(logging.DEBUG):
+                    _LOG.debug("Run: %s", list2cmdline(task))
+
+                with open(stdin) as _in, open(logfile('stdout'), "wb") as _out, open(logfile('stderr'), "wb") as _err:
+                    processes.append([Popen(task, stdin=_in, stdout=_out, stderr=_err), node.host, process_id])
+
+            # check for ended processes
+            for p in processes:
+                [process, host, p_id] = p
```

Comment:
brackets are not needed here, `process, host, p_id = p` works exactly the same and I think most people used more to this syntax